src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+    /// whatever level in the tree is visited. This has the effect of comprehehsively 
+    /// 'hydrating' all possible SARIF constructs as defined by the C# OM.
+    /// </summary>
+    public class DefaultObjectPopulatingVisitor : SarifRewritingVisitor
+    {
+        public delegate object PrimitiveValueBuilder(bool isRequired);
+
+        // The default builders create a placeholder value for all required property primitives
+        // and emits default (null, 0, false, etc.) values for the remainder. As a result, a
+        // default hydration of this class should contain no non-default values for 
+        // JSON primitive-valued properties that aren't required
+        public static IDictionary<Type, PrimitiveValueBuilder> GetBuildersForRequiredPrimitives()
+        {
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return isRequired; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 0; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 0; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "Required string value." : null; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.contoso.com") : null; } }
+            };
+        }
+
+        // These builders exhaustively populate the generated SARIF file, including non-required JSON primitives
+        public static IDictionary<Type, PrimitiveValueBuilder> GetBuildersForAllPrimitives()
+        {
+            // The values in this output are designed to allow conformance to the schema. In our schema,.
+            // for example, some integer values must be > 0. And so, our optional integer value is 
+            // written as 1.
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return true; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 1; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 1; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "[Required string value]" : "[Optional string value]"; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(1776, 7, 4).ToUniversalTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.value.contoso.com") : new Uri("https://optional.uri.value.contoso.com"); } }
+            };
+        }
+
+        public DefaultObjectPopulatingVisitor(JsonSchema schema, IDictionary<Type, PrimitiveValueBuilder> primitiveValueBuilders = null)
```

Comment:
>schema [](start = 57, length = 6)

Note to self: It is unusual for a visitor to take the schema, since its generated code incorporates knowledge of the schema. But @michaelcfanning has told me that this visitor engages in some hijinks with the schema, so I will wait and see...

Reply:
Ok, now I know what it's for. It isn't actually needed; see comment below about reflecting on `DataMemberAttribute`.

---
In reply to: [229125407](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229125407) [](ancestors = 229125407)